account_invoice_ubl_mass_download/models/account_invoice.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2021 Coop IT Easy (https://coopiteasy.be)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import base64
+import io
+import logging
+import zipfile
+
+from odoo import api, models
+
+_logger = logging.getLogger(__name__)
+
+ATTACHMENT_TMP_NAME = "temporary-invoice-ubl-zip-file"
+
+
+class AccountInvoice(models.Model):
+
+    _inherit = "account.invoice"
+
+    @api.multi
+    def zip_ubl_xml_and_pdf(self):
+        """
+        Create a zip file with XML UBL and PDF for each invoice.
+        """
+        with io.BytesIO() as buffer:
+            with zipfile.ZipFile(
+                buffer, mode="w", compression=zipfile.ZIP_DEFLATED
+            ) as ziparc:
+                for invoice in self:
+                    assert invoice.type in ("out_invoice", "out_refund")
+                    assert invoice.state in ("open", "paid")
+                    # XML UBL
+                    version = invoice.get_ubl_version()
+                    xml_string = invoice.generate_ubl_xml_string(
+                        version=version
+                    )
+                    xmlname = "{}-{}".format(
+                        invoice.number.replace("/", "-"),
+                        invoice.get_ubl_filename(version=version),
+                    )
+                    ziparc.writestr(xmlname, xml_string)
+                    # PDF
+                    reportinv = (
+                        self.env["ir.actions.report"]
+                        ._get_report_from_name("account.report_invoice")
+                        .render_qweb_pdf([invoice.id])
+                    )
+                    invpdf = reportinv[0]
+                    pdfname = "{}.pdf".format(
+                        invoice.number.replace("/", "-"),
+                    )
+                    ziparc.writestr(pdfname, invpdf)
+            ctx = {}
+            zipname = "XML-UBL-PDF.zip"
```

Comment:
Ought this contain a date or nah?

Reply:
I can include the date in the `zipname` but I'm not sure about the benefice of that. The user will certainly extract de zip file to get all the individual files that contains the invoice number. Can you explain more what's the benefice ? Thanks. :)

Reply:
There's no real benefit as such, and it needn't even be a date specifically; I just think it's nice when there's a little extra context in the filename. This has two helpful effects:

- When you (much) later encounter the file again, you have a tiny aid in figuring out what the file was about.
- When you download this zip file twice, there's a meaningful way to distinguish between them.

But it's really not a huge blocker.

Reply:
As, there where time left on the task, I added a timestamp on the zipname and moved the constant at the begin of the file as suggested. :) Thanks.